pkg/webhook/elasticquota/quota_topology.go

```diff
@@ -0,0 +1,664 @@
+					return false
+				}
+			}
+		} else {
+			klog.Errorf("BUG quotaInfoMap and quotaTree information out of sync, losed :%v", childName)
+			return false
+		}
+	}
+	return true
+}
+
+func (qt *quotaTopology) checkQuotaHierarchyHasChange(oldQuotaInfo, newQuotaInfo *core.QuotaInfo) error {
+	// parentQuota -> childQuota, check whether the quotaInfo has child or not
+	if oldQuotaInfo.IsParent == true && newQuotaInfo.IsParent == false {
+		childQuotaSet, exist := qt.quotaHierarchyInfo[newQuotaInfo.Name]
+		if !exist {
+			return fmt.Errorf(":%v not hound quota hierarchy", newQuotaInfo.Name)
+		}
+		if len(childQuotaSet) > 0 {
+			return fmt.Errorf("can not modify IsParent to false, because this groups still has child quota")
+		}
+	}
+
+	if oldQuotaInfo.ParentName != newQuotaInfo.ParentName {
+		if oldQuotaInfo.ParentName != extension.RootQuotaName {
+			delete(qt.quotaHierarchyInfo[oldQuotaInfo.ParentName], oldQuotaInfo.Name)
+		}
+
+		if newQuotaInfo.ParentName != extension.RootQuotaName {
+			qt.quotaHierarchyInfo[newQuotaInfo.ParentName][newQuotaInfo.Name] = struct{}{}
+		}
+	}
+	return nil
+}
+
+func toElasticQuota(obj interface{}) *v1alpha1.ElasticQuota {
+	if obj == nil {
+		return nil
+	}
+
+	var unstructuredObj *unstructured.Unstructured
+	switch t := obj.(type) {
+	case *v1alpha1.ElasticQuota:
+		return obj.(*v1alpha1.ElasticQuota)
+	case *unstructured.Unstructured:
+		unstructuredObj = obj.(*unstructured.Unstructured)
+	case cache.DeletedFinalStateUnknown:
+		var ok bool
+		unstructuredObj, ok = t.Obj.(*unstructured.Unstructured)
+		if !ok {
+			klog.Errorf("Fail to convert quota object %T to *unstructured.Unstructured", obj)
+			return nil
+		}
+	default:
+		klog.Errorf("Unable to handle quota object in %T", obj)
+		return nil
+	}
+
+	quota := &v1alpha1.ElasticQuota{}
+	err := scheme.Scheme.Convert(unstructuredObj, quota, nil)
+	if err != nil {
+		klog.Errorf("Fail to convert unstructed object %v to Quota: %v", obj, err)
+		return nil
+	}
+	return quota
+}
+
+func quotaFieldsCopy(q *v1alpha1.ElasticQuota) v1alpha1.ElasticQuota {
+	return v1alpha1.ElasticQuota{
+		ObjectMeta: metav1.ObjectMeta{
+			Labels: map[string]string{
+				extension.LabelQuotaParent:       q.Labels[extension.LabelQuotaParent],
+				extension.LabelQuotaIsParent:     q.Labels[extension.LabelQuotaIsParent],
+				extension.LabelAllowLentResource: q.Labels[extension.LabelAllowLentResource],
+			},
+			Annotations: map[string]string{
+				extension.AnnotationSharedWeight: q.Annotations[extension.AnnotationSharedWeight],
+			},
+		},
+		Spec: *q.Spec.DeepCopy(),
+	}
+}
+
+func (qt *quotaTopology) getQuotaNameFromPodNoLock(pod *v1.Pod) string {
+	quotaLabelName := pod.Labels[extension.LabelQuotaName]
+	if quotaLabelName == "" {
+		quotaList := &v1alpha1.ElasticQuotaList{}
+		opts := &client.ListOptions{
+			LabelSelector: labels.Everything(),
+		}
+		err := qt.client.List(context.TODO(), quotaList, opts, utilclient.DisableDeepCopy)
```

Comment:
why this implementation is different with plugin. besides what's the meanning of "listAll quotas and pick first"? at least use namespace???

Reply:
fixed...